lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/TaskHostBase.cs

```diff
@@ -0,0 +1,242 @@
+        /// </summary>
+        /// <param name="groupCommunicationsClient"></param>
+        /// <param name="taskCloseCoordinator"></param>
+        /// <param name="invokeGc"></param>
+        protected TaskHostBase(
+            IGroupCommClient groupCommunicationsClient,
+            TaskCloseCoordinator taskCloseCoordinator,
+            bool invokeGc)
+        {
+            _groupCommunicationsClient = groupCommunicationsClient;
+            _communicationGroupClient = groupCommunicationsClient.GetCommunicationGroup(IMRUConstants.CommunicationGroupName);
+
+            InvokeGc = invokeGc;
+            _taskCloseCoordinator = taskCloseCoordinator;
+            CancellationSource = new CancellationTokenSource();
+        }
+
+        /// <summary>
+        /// Handle the exceptions in the Call() method
+        /// Default to IMRUSystemException to make it recoverable
+        /// </summary>
+        /// <param name="memento"></param>
+        /// <returns></returns>
+        public byte[] Call(byte[] memento)
+        {
+            Logger.Log(Level.Info, "Entering {0} Call().", TaskHostName);
+            try
+            {
+                return TaskBody(memento);
+            }
+            catch (OperationCanceledException e)
+            {
+                HanldeCommunicationException(e);
+            }
+            catch (IOException e)
+            {
+                HanldeCommunicationException(e);
+            }
+            catch (TcpClientConnectionException e)
+            {
+                HanldeCommunicationException(e);
+            }
+            catch (RemotingException e)
+            {
+                HanldeCommunicationException(e);
+            }
+            catch (SocketException e)
+            {
+                HanldeCommunicationException(e);
+            }
+            catch (AggregateException e)
+            {
+                if (e.InnerException != null && e.InnerException is IOException)
+                {
+                    HanldeCommunicationException(e);
+                }
+                else
+                {
+                    HanldeSystemException(e);
+                }
+            }
+            catch (Exception e)
+            {
+                HanldeSystemException(e);
+            }
+            finally
+            {
+                FinallyBlock();
+                _taskCloseCoordinator.SignalTaskStopped();
+            }
+            Logger.Log(Level.Info, "{0} returned with cancellation token:{0}.", TaskHostName, CancellationSource.IsCancellationRequested);
+            return null;
+        }
+
+        /// <summary>
+        /// The body of Call method. Subclass must override it. 
+        /// </summary>
+        /// <param name="memento"></param>
+        /// <returns></returns>
+        protected abstract byte[] TaskBody(byte[] memento);
+        
+        /// <summary>
+        /// The code that needs to be executed in finally block. 
```

Comment:
I'd say "Code that needs to be executed after an exception occurs" (finally block is implementation detail, while documentation should be higher level)